balloc/balloc.c

```diff
@@ -2979,29 +3069,61 @@ static const struct m0_ad_balloc_ops balloc_ops = {
 };
 
 static int balloc_trees_create(struct m0_balloc    *bal,
+			       struct m0_be_seg    *seg,
 			       struct m0_be_tx     *tx,
-			       const struct m0_fid *fid)
+			       const struct m0_fid *bfid)
 {
 	int rc;
+	struct m0_btree_type  bt;
+	struct m0_btree_op    b_op = {};
+	struct m0_fid         fid;
 
-	rc = M0_BE_OP_SYNC_RET(op,
-		       m0_be_btree_create(&bal->cb_db_group_extents, tx, &op,
-				  &M0_FID_TINIT('b',
-						M0_BBT_BALLOC_GROUP_EXTENTS,
-						fid->f_key)),
-		       bo_u.u_btree.t_rc);
-	if (rc != 0)
+	M0_ALLOC_PTR(bal->cb_db_group_extents);
+	if (bal->cb_db_group_extents == NULL)
+		return M0_ERR(-ENOMEM);
+
+	M0_ALLOC_PTR(bal->cb_db_group_desc);
+	if (bal->cb_db_group_desc == NULL) {
+		m0_free0(&bal->cb_db_group_extents);
+		return M0_ERR(-ENOMEM);
+	}
+
+	bt = (struct m0_btree_type){.tt_id = M0_BT_BALLOC_GROUP_EXTENTS,
+		.ksize = M0_MEMBER_SIZE(struct m0_ext, e_start),
+		.vsize = M0_MEMBER_SIZE(struct m0_ext, e_end),
+	};
+	fid = M0_FID_TINIT('b', M0_BBT_BALLOC_GROUP_EXTENTS, bfid->f_key);
+	rc = M0_BTREE_OP_SYNC_WITH_RC(&b_op,
+				      m0_btree_create(bal->cb_ge_node,
+						      sizeof bal->cb_ge_node,
+						      &bt, &b_op,
+						      bal->cb_db_group_extents,
+						      seg, &fid, tx));
+	if (rc != 0) {
+		m0_free0(&bal->cb_db_group_extents);
+		m0_free0(&bal->cb_db_group_desc);
 		return M0_ERR(rc);
+	}
 
-	rc = M0_BE_OP_SYNC_RET(op,
-		       m0_be_btree_create(&bal->cb_db_group_desc, tx, &op,
-				  &M0_FID_TINIT('b', M0_BBT_BALLOC_GROUP_DESC,
-						fid->f_key)),
-		       bo_u.u_btree.t_rc);
+	bt = (struct m0_btree_type){.tt_id = M0_BT_BALLOC_GROUP_DESC,
+		.ksize = M0_MEMBER_SIZE(struct m0_balloc_group_desc,
+					bgd_groupno),
+		.vsize = sizeof(struct m0_balloc_group_desc),
+	};
+	fid = M0_FID_TINIT('b', M0_BBT_BALLOC_GROUP_DESC, bfid->f_key);
+	rc = M0_BTREE_OP_SYNC_WITH_RC(&b_op,
+				      m0_btree_create(bal->cb_gd_node,
+						      sizeof bal->cb_gd_node,
+						      &bt, &b_op,
+						      bal->cb_db_group_desc,
+						      seg, &fid, tx));
 
 	if (rc != 0) {
-		M0_BE_OP_SYNC(op,
-		      m0_be_btree_destroy(&bal->cb_db_group_extents, tx, &op));
+		M0_BTREE_OP_SYNC_WITH_RC(&b_op,
+				m0_btree_destroy(bal->cb_db_group_extents,
```

Comment:
alignment

Reply:
Yes, that was intentional as all other options were resulting into exceeding 80 char limit. If its okay, please ignore this one.